### PR TITLE
execgen: try not to touch .eg.go files

### PR DESCRIPTION
**Pull request by jordanlewis**
This changes the Makefile to output .eg.go output to a tmp file before
actually editing .eg.go files, to prevent build churn.

Release note: None

### CODE REVIEW

**Comment by jordanlewis** [comment]
I tried this out but it doesn't seem to work - if I `rm bin/execgen` and rebuild, even though `cmp` returns 0, the build still has to do `colexec` for some reason.

**Comment by jordanlewis** [comment]
Another problem is that this always runs the code that regenerates the `.tmp` files for some reason. My make fu is not up to this task

**Review by petermattis — COMMENTED** [review]
> Another problem is that this always runs the code that regenerates the .tmp files for some reason. My make fu is not up to this task

This is part of the problem with Make. You can get it to do anything, but it takes expertise. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45737#-:-M1dxHtI29JfSmY-OIcO:b-3eoj1r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, and @petermattis)

___
*[Makefile, line 1539 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45737#-M1dxJx7BpSgR5R_CQiF:-M1dxJx7BpSgR5R_CQiG:b-t0z6ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/d374874a9e9efefad7a686f5024e378fc0930aa0/Makefile#L1539)):*
> ```Makefile
> 
> $(EXECGEN_TARGETS): bin/execgen
> 	execgen $@ > $@.tmp
> ```

It is a bit odd that execgen takes the output file as input. It should take the input file as input, and then write the output to stdout. I'd then have to dig up the make syntax for finding the first dependency. Something like: `execgen $^ > $@`. For now, I'd be fine if you changed this to `execgen $@ > $@` and removed the next line since it isn't doing the right thing anyways for some reason. I can then take a look after this PR is merged about the correct make incantations here.

**Comment by jordanlewis** [comment]
> I tried this out but it doesn't seem to work - if I rm bin/execgen and rebuild, even though cmp returns 0, the build still has to do colexec for some reason.

I don't understand this at all, because I double checked that the rules I added don't edit the mtime of any of the files in `colexec`...

> It is a bit odd that execgen takes the output file as input. It should take the input file as input, and then write the output to stdout. 

execgen doesn't really work like this, since in some cases one input file can make more than one output file. maybe we could coerce it to behave more normally, but I don't think it really matters.

> removed the next line since it isn't doing the right thing anyways for some reason. I can then take a look after this PR is merged about the correct make incantations here.

I think the next line does actually do the right thing. We need to figure out why Go is getting triggered to rebuild.

File: pkg/sql/colexec/execgen/cmd/execgen/main.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func (g *execgen) run(args ...string) bool {
 			g.reportError(errors.Errorf("unrecognized filename: %s", file))
 			return false
 		}
-		if err := g.generate(gen, out); err != nil {
+		if err := g.generate(gen); err != nil {
```
Comment (irfansharif) [inline pkg/sql/colexec/execgen/cmd/execgen/main.go]: Perhaps rename `out`?

**Review by irfansharif — COMMENTED** [review]
Taking a (brief) look at what the right Makefile diff should be.

**Comment by irfansharif** [comment]
Jordan, it seems that `rowstovec.eg.go` is the problem.

```
$ execgen pkg/sql/colexec/rowstovec.eg.go > /tmp/a
$ execgen pkg/sql/colexec/rowstovec.eg.go > /tmp/b
$ execgen pkg/sql/colexec/rowstovec.eg.go > /tmp/c
$ execgen pkg/sql/colexec/rowstovec.eg.go > /tmp/d
$ cmp /tmp/a /tmp/b
/tmp/a /tmp/b differ: char 1319, line 41
$ cmp /tmp/a /tmp/c
/tmp/a /tmp/c differ: char 1319, line 41
$ cmp /tmp/a /tmp/d
/tmp/a /tmp/d differ: char 1887, line 64
```

**Comment by irfansharif** [comment]
```diff
$ diff /tmp/a /tmp/b  | head -n 20
41,42c41,42
<                       case types.StringFamily:
<                               col := vec.Bytes()
---
>                       case types.DateFamily:
>                               col := vec.Int64()
60,133c60,61
<                                               castV := v.([]byte)
<                                               col.Set(i, castV)
<                                       }
<                               }
<                       case types.IntFamily:
<                               switch columnType.Width() {
<                               case 16:
<                                       col := vec.Int16()
<                                       datumToPhysicalFn := typeconv.GetDatumToPhysicalFn(columnType)
<                                       for i := range rows {
<                                               row := rows[i]
<                                               if row[columnIdx].Datum == nil {
<                                                       if err = row[columnIdx].EnsureDecoded(columnType, alloc); err != nil {
```

The generated code doesn't deterministically sort the `switch/case` listings.

**Comment by jordanlewis** [comment]
That's annoying, thanks for catching it...

**Comment by jordanlewis** [comment]
Ok, I fixed it and verified that it works now. Realistically I don't think it's gonna speed up the build that much because `colexec` itself compiles very quickly. But we'll see!

File: Makefile [hunk 2]
```diff
@@ -1536,7 +1536,8 @@ pkg/sql/colexec/vec_comparators.eg.go: pkg/sql/colexec/vec_comparators_tmpl.go
 pkg/sql/colexec/window_peer_grouper.eg.go: pkg/sql/colexec/window_peer_grouper_tmpl.go
 
 $(EXECGEN_TARGETS): bin/execgen
-	execgen $@
+	execgen $@ > $@.tmp; cmp $@.tmp $@ && rm -f $@.tmp || mv $@.tmp $@
```
Comment (irfansharif) [inline Makefile]: Splitting it into two lines is more "idiomatic".

**Review by irfansharif — APPROVED** [review]
LGTM, thanks for fixing it all. I'll rebase #45611 on top of this once in and see where we're losing the time, if the code generation itself takes a few seconds I'll keep it as is and skip it altogether (though this change is still a step in the right direction).

File: Makefile (hunk 2 above)
Comment (jordanlewis) [reply Makefile]: Yeah but it's super noisy and gross in the Makefile output so I'll avoid unless you insist.

**Pull request merged**